### PR TITLE
Add redundancy

### PR DESCRIPTION
**Pull request by Rdk-T**
close https://github.com/Siviuze/KickCAT/issues/38

This PR add the redundancy support to kickcat master:

- rework of the link object
- Update the corresponding unit tests and add redundancy testing
- default behavior is now to always use redundancy, possibility to pass a dummy interface to the link in case of a no redundancy system

For the record, I noticed a buggy behavior in the mock used in the test when trying to write and read multiple datagrams with the mocked functions `checkSendFrame` and `handleReply`

### CODE REVIEW

**Comment by Siviuze** [comment]
Don't forget to update the README :)

File: README.md [hunk 1]
```diff
@@ -2,9 +2,12 @@
 
 Kick-start your slaves!
 
+Thin EtherCAT stack desgined to be embedded in a more complex software with with efficiency in mind.
```
Comment (Siviuze) [inline README.md:5]: typo (with with)

File: examples/easycat/easycat_example.cc [hunk 2]
```diff
@@ -17,14 +19,51 @@ using namespace kickcat;
...
-    Bus bus(socket);
+
+    auto reportRedundancy = []()
+    {
+        printf("Redundancy has been activated due to loss of a cable \n");
+    };
+
+    auto socketNominal = std::make_shared<Socket>();
+
+    std::shared_ptr<AbstractSocket> socketRedundancy;
```
Comment (Siviuze) [inline examples/easycat/easycat_example.cc:37]: Parse all command line arguments in one section and then use the result instead of mixing parsing and bus creation.

File: examples/easycat/easycat_example.cc [hunk 3]
```diff
@@ -89,11 +128,11 @@ int main(int argc, char* argv[])
         return 1;
     }
 
-    socket->setTimeout(500us);
+    microseconds nominal_timeout = 500us;
+    socketNominal->setTimeout(nominal_timeout);
+    socketRedundancy->setTimeout(nominal_timeout);
 
     constexpr int64_t LOOP_NUMBER = 12 * 3600 * 1000; // 12h
-    FILE* stat_file = fopen("stats.csv", "w");
```
Comment (Siviuze) [inline examples/easycat/easycat_example.cc:95]: Why did you remove the stats generator?

File: include/kickcat/Frame.h [hunk 4]
```diff
@@ -28,15 +28,16 @@ namespace kickcat
         /// \return true if full after adding datagram, false otherwise
         void addDatagram(uint8_t index, enum Command command, uint32_t address, void const* data, uint16_t data_size);
 
-        void clear(); // reset frame context
+        void resetContext();
```
Comment (Siviuze) [inline include/kickcat/Frame.h:31]: Add a comment of what this call is doing (seems even more important now to differentiate clear() and resetContext())

File: include/kickcat/Frame.h [hunk 5]
```diff
@@ -28,15 +28,16 @@ namespace kickcat
...
+        void clear();
```
Comment (Siviuze) [inline include/kickcat/Frame.h:32]: same

File: include/kickcat/Frame.h [hunk 6]
```diff
@@ -49,24 +50,31 @@ namespace kickcat
 
         /// \return true if datagram can be extracted, false otherwise
         bool isDatagramAvailable() const { return is_datagram_available_; }
+        void setIsDatagramAvailable() { is_datagram_available_ = true; }
 
-        // handle bus access
-        void read(std::shared_ptr<AbstractSocket> socket);
-        void write(std::shared_ptr<AbstractSocket> socket);
+        void setSourceMAC(mac const src_mac);
```
Comment (Siviuze) [inline include/kickcat/Frame.h:55]: const&

File: include/kickcat/Frame.h [hunk 7]
```diff
@@ -49,24 +50,31 @@ namespace kickcat
...
 
         // helper to access raw frame (mostly for unit testing)
         uint8_t* data() { return frame_.data(); }
+        EthercatHeader* header()   { return header_; }
+        EthernetHeader* ethernet() { return ethernet_; }
 
+
+
+        /// For writeThenRead
+        uint8_t* last_datagram_{nullptr};   // Last **written** datagram
```
Comment (Siviuze) [inline include/kickcat/Frame.h:65]: Why do you need direct access to member? wirteThenRead() doesn't seem to need tsuch access

File: include/kickcat/Link.h [hunk 8]
```diff
@@ -1,5 +1,5 @@
-#ifndef KICKCAT_LINK_H
-#define KICKCAT_LINK_H
+#ifndef KICKCAT_LINK_REDUNDANCY_H
```
Comment (Siviuze) [inline include/kickcat/Link.h:1]: Keep old header guard since the filename didn't evole

File: include/kickcat/Link.h [hunk 9]
```diff
@@ -12,18 +12,21 @@ namespace kickcat
 {
     class AbstractSocket;
 
-    /// \brief Handle link layer
```
Comment (Siviuze) [inline include/kickcat/Link.h:15]: Keep this comment and update it if needed

File: include/kickcat/Link.h [hunk 10]
```diff
@@ -12,18 +12,21 @@ namespace kickcat
...
+                       mac const src_mac_nominal = PRIMARY_IF_MAC,
+                       mac const src_mac_redundancy = SECONDARY_IF_MAC);
         ~Link() = default;
 
-        /// \brief helper for trivial access (i.e. most of the init bus frames)
-        void writeThenRead(Frame& frame);
+        /// \brief   Helper for trivial access (i.e. most of the init bus frames)
+        ///
+        /// \details Since this method is only used for non real time operation, the redundancy mechanism used is slower
+        ///          but guaranty slave order access, ie for setAdresses.
```
Comment (Siviuze) [inline include/kickcat/Link.h:28]: typo: setAddresses()

File: include/kickcat/SocketNull.h [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <string>
+#include <chrono>
+
+#include "Error.h"
+#include "AbstractSocket.h"
+
+namespace kickcat
+{
+    using namespace std::chrono;
```
Comment (Siviuze) [inline include/kickcat/SocketNull.h:12]: This should not be needed (already done in AbstractSocket.h)

File: include/kickcat/SocketNull.h [hunk 12]
```diff
@@ -0,0 +1,29 @@
+#ifndef KICKCAT_VOID_SOCKET_H
```
Comment (Siviuze) [inline include/kickcat/SocketNull.h:1]: Use file name as reference for header guards

File: include/kickcat/protocol.h [hunk 13]
```diff
@@ -10,11 +10,15 @@
 
 namespace kickcat
 {
+    constexpr int32_t  MAC_SIZE = 6;
+
+    using mac = uint8_t[MAC_SIZE];
```
Comment (Siviuze) [inline include/kickcat/protocol.h:15]: I wonder if capital letters is more readable for MAC type :thinking:

File: include/kickcat/protocol.h [hunk 14]
```diff
@@ -10,11 +10,15 @@
...
+
     // Ethernet description
     struct EthernetHeader
     {
-        uint8_t dst_mac[6];
-        uint8_t src_mac[6];
+        mac dst_mac;
```
Comment (Siviuze) [inline include/kickcat/protocol.h:20]: Seems redundant to postfix _mac since the type define it.

File: src/Bus.cc [hunk 15]
```diff
@@ -430,7 +434,6 @@ namespace kickcat
             {
                 if (wkc != pi_frame.inputs.size())
                 {
-                    DEBUG_PRINT("Invalid working counter\n");
```
Comment (Siviuze) [inline src/Bus.cc:433]: No debug_print here anymore?

**Comment by Siviuze** [comment]
There is a dangling file: [src/Bus.cc.orig](https://github.com/Siviuze/KickCAT/pull/43/files#diff-682bda9af24cace77ea7f37fc62d90c3345620642b66315fe2e4d81f02512663)

File: src/Link.cc [hunk 16]
```diff
@@ -120,16 +93,215 @@ namespace kickcat
...
+        frame_redundancy_.clear();
+        frame_redundancy_.resetContext();
+
+        if (written < 0)
+        {
+            is_frame_sent_redundancy = false;
+            DEBUG_PRINT("Redundancy: write failed\n");
+        }
+
+        else if (written != toWrite)
```
Comment (Siviuze) [inline src/Link.cc:230]: May be done with only this condition and written added to the debug trace

File: src/Link.cc [hunk 17]
```diff
@@ -120,16 +93,215 @@ namespace kickcat
...
+        bool is_frame_sent_nominal = true;
+        frame_nominal_.setSourceMAC(PRIMARY_IF_MAC);
+        int32_t toWrite = frame_nominal_.finalize();
+        int32_t written = socket_nominal_->write(frame_nominal_.data(), toWrite);
+        if (written < 0)
+        {
+            is_frame_sent_nominal = false;
+            DEBUG_PRINT("Nominal: write failed\n");
+        }
+        else if (written != toWrite)
```
Comment (Siviuze) [inline src/Link.cc:211]: May be done with only this condition and written added to the debug trace

File: src/Link.cc [hunk 18]
```diff
@@ -120,16 +93,215 @@ namespace kickcat
...
+        {
+            is_frame_sent_nominal = false;
+            DEBUG_PRINT("Nominal: Wrong number of bytes written");
+        }
+
+        bool is_frame_sent_redundancy = true;
+        frame_nominal_.setSourceMAC(SECONDARY_IF_MAC);
+        written = socket_redundancy_->write(frame_nominal_.data(), toWrite);
```
Comment (Siviuze) [inline src/Link.cc:219]: Seems possible to use a lambda to factorise the code

File: src/Link.cc [hunk 19]
```diff
@@ -120,16 +93,215 @@ namespace kickcat
...
+
+        else if (written != toWrite)
+        {
+            is_frame_sent_redundancy = false;
+            DEBUG_PRINT("Redundancy: Wrong number of bytes written");
+        }
+
+        if (is_frame_sent_nominal or is_frame_sent_redundancy)
+        {
+            sent_frame_++;
```
Comment (Siviuze) [inline src/Link.cc:238]: Better use prefix increment (++sent_frame_) instead since you'll never need the old value.

File: src/Link.cc [hunk 20]
```diff
@@ -120,16 +93,215 @@ namespace kickcat
...
+
+    std::tuple<DatagramHeader const*, uint8_t*, uint16_t> Link::nextDatagram()
+    {
+        bool nom = frame_nominal_.isDatagramAvailable();
+        bool red = frame_redundancy_.isDatagramAvailable();
+
+        auto [header_nominal, data_nominal, wkc_nominal] = frame_nominal_.nextDatagram();
+        auto [header_redundancy, data_redundancy, wkc_redundancy] = frame_redundancy_.nextDatagram();
+
+        // When using the bus without redundancy not nom and red is the used case.
```
Comment (Siviuze) [inline src/Link.cc:291]: Why is this the case without redundancy? Isn't red the one without datagrams in this case?

File: unit/Mocks.h [hunk 21]
```diff
@@ -1,9 +1,35 @@
...
+    };
+
+    template<typename T>
+    struct DatagramCheck
+    {
+        Command cmd;
+        T to_check;
+        bool check_payload{true};
+        DatagramCheck(Command cmd_, T to_check_, bool check_payload_ = true)
+        : cmd(cmd_)
```
Comment (Siviuze) [inline unit/Mocks.h:26]: indent

File: unit/link-t.cc [hunk 22]
```diff
@@ -1,5 +1,6 @@
 #include <gtest/gtest.h>
 #include <cstring>
+#include <string.h>
```
Comment (Siviuze) [inline unit/link-t.cc:3]: redundant with cstring

File: examples/easycat/easycat_example.cc (hunk 3 above)
Comment (Rdk-T) [reply examples/easycat/easycat_example.cc:95]: By mistake

File: CMakeLists.txt [hunk 23]
```diff
@@ -80,7 +80,7 @@ if (GTest_FOUND)
     setup_target_for_coverage_gcovr_html(
         NAME coverage
         EXECUTABLE kickcat_unit
-        EXCLUDE "unit/*" ".*gtest.*" "example" ".*gmock.*" ".*LinuxSocket*" "src/Prints.cc"
+        EXCLUDE "unit/*" ".*gtest.*" "example" ".*gmock.*" ".*/OS/.*"
         )
 
     setup_target_for_coverage_gcovr_xml(
```
Comment (Siviuze) [inline CMakeLists.txt:86]: update exclude list here also

File: include/kickcat/Frame.h (hunk 7 above)
Comment (Rdk-T) [reply include/kickcat/Frame.h:65]: I don't need it anymore

File: src/Link.cc (hunk 19 above)
Comment (Rdk-T) [reply src/Link.cc:238]: Why is it better ?

File: src/Link.cc (hunk 20 above)
Comment (Rdk-T) [reply src/Link.cc:291]: There is some ambiguity with the names frame_redundancy and frame_nominal.

When redundancy is working properly:
- nominal_frame is written on nominal_socket and read on redundancy_socket
- redundancy_frame is written on redundancy_socket and read on nominal socket
So the two frames are stored as member of Link under these names (point of view of the writing).

When there is no redundancy interface:
- nominal_frame is written on nominal_socket
- when reading redundancy socket there is nothing for nominal_frame (not connected)
- when reading nominal_socket the frame is stored in redundancy_frame.

Comment (Siviuze) [reply src/Link.cc:279]: got it

File: src/Link.cc (hunk 19 above)
Comment (Siviuze) [reply src/Link.cc]: i = 2
b = ++i
a = i++

=> a == b == 3
By convention, I prefer to use prefix increment when the value before increment is of no use.

**Review by Siviuze — APPROVED** [review]
